service/matching/reachability.go

```diff
@@ -129,56 +130,26 @@ func (rc *reachabilityCalculator) run(ctx context.Context, buildId string) (enum
 
 // getBuildIdsOfInterest returns a list of build ids that point to the given buildId in the graph
 // of redirect rules and adds the given build id to that list.
-// It considers rules if the deletion time is within versioningReachabilityDeletedRuleInclusionPeriod.
+// It considers rules if the deletion time is nil or within the given deletedRuleInclusionPeriod.
 func (rc *reachabilityCalculator) getBuildIdsOfInterest(
 	buildId string,
-	includeRecentlyDeleted bool) []string {
-	return append(rc.getUpstreamHelper(buildId, includeRecentlyDeleted, nil), buildId)
-}
-
-func (rc *reachabilityCalculator) getUpstreamHelper(
-	buildId string,
-	includeRecentlyDeleted bool,
-	visited []string) []string {
-	var upstream []string
-	visited = append(visited, buildId)
-	directSources := rc.getSourcesForTarget(buildId, includeRecentlyDeleted)
+	deletedRuleInclusionPeriod *time.Duration) []string {
 
-	for _, src := range directSources {
-		if !slices.Contains(visited, src) {
-			upstream = append(upstream, src)
-			upstream = append(upstream, rc.getUpstreamHelper(src, includeRecentlyDeleted, visited)...)
+	withinRuleInclusionPeriod := func(clk *clock.HybridLogicalClock) bool {
+		if clk == nil {
+			return true
 		}
-	}
-
-	// dedupe
-	upstreamUnique := make(map[string]bool)
-	for _, bid := range upstream {
-		upstreamUnique[bid] = true
-	}
-	upstream = make([]string, 0)
-	for k := range upstreamUnique {
-		upstream = append(upstream, k)
-	}
-	return upstream
-}
-
-// getSourcesForTarget gets the first-degree sources for any redirect rule targeting buildId
-func (rc *reachabilityCalculator) getSourcesForTarget(buildId string, includeRecentlyDeleted bool) []string {
-	var sources []string
-	for _, rr := range rc.redirectRules {
-		if rr.GetRule().GetTargetBuildId() == buildId &&
-			((!includeRecentlyDeleted && rr.GetDeleteTimestamp() == nil) ||
-				(includeRecentlyDeleted && rc.withinDeletedRuleInclusionPeriod(rr.GetDeleteTimestamp()))) {
-			sources = append(sources, rr.GetRule().GetSourceBuildId())
+		if deletedRuleInclusionPeriod == nil {
+			return false
 		}
+		return hlc.Since(clk) <= *deletedRuleInclusionPeriod
 	}
 
-	return sources
-}
+	includedRules := util.FilterSlice(slices.Clone(rc.redirectRules), func(rr *persistencespb.RedirectRule) bool {
```

Comment:
I really wish we'd broken this out into, say, `common/collections/slice` so this could be `slice.Filter()` (and also `slice.Fold`, etc).

It's unfortunate that the official `slices` package doesn't contain these